src/plots/grid.js

```diff
@@ -0,0 +1,409 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Lib = require('../lib');
+var domainAttrs = require('./domain').attributes;
+var counterRegex = require('../lib/regex').counter;
+var cartesianIdRegex = require('./cartesian/constants').idRegex;
+
+
+var gridAttrs = exports.attributes = {
+    rows: {
+        valType: 'integer',
+        min: 1,
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'The number of rows in the grid. If you provide a 2D `subplots`',
+            'array or a `yaxes` array, its length is used as the default.',
+            'But it\'s also possible to have a different length, if you',
+            'want to leave a row at the end for non-cartesian subplots.'
+        ].join(' ')
+    },
+    roworder: {
+        valType: 'enumerated',
+        values: ['top to bottom', 'bottom to top'],
+        dflt: 'top to bottom',
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Is the first row the top or the bottom? Note that columns',
+            'are always enumerated from left to right.'
+        ].join(' ')
+    },
+    columns: {
+        valType: 'integer',
+        min: 1,
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'The number of columns in the grid. If you provide a 2D `subplots`',
+            'array, the length of its longest row is used as the default.',
+            'If you give an `xaxes` array, its length is used as the default.',
+            'But it\'s also possible to have a different length, if you',
+            'want to leave a row at the end for non-cartesian subplots.'
+        ].join(' ')
+    },
+    subplots: {
+        valType: 'info_array',
+        freeLength: true,
+        dimensions: 2,
+        items: {valType: 'enumerated', values: [counterRegex('xy').toString(), '']},
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Used for freeform grids, where some axes may be shared across subplots',
+            'but others are not. Each entry should be a cartesian subplot id, like',
+            '*xy* or *x3y2*, or ** to leave that cell empty. You may reuse x axes',
+            'within the same column, and y axes within the same row.',
+            'Non-cartesian subplots and traces that support `domain` can place themselves',
+            'in this grid separately using the `gridcell` attribute.'
```

Comment:
There would be something nice about specifying non-cartesian subplots in this array as well... but there are two reasons I didn't allow that:
- It's easy to specify `'scene2'` and `'geo3'` etc but how do you specify a trace (like pie)? You'd have to use `uid` or something. I feel like unless we can get everything, we should only allow cartesian subplots this way.
- What if you use the `xaxes` and `yaxes` form (or one of the defaults with `pattern`) but you still want to insert something else? Like, say you have a splom but you want to put pies along the diagonal, or a splom showing only above the diagonal and you want to put some unrelated subplots below the diagonal?

Also I should mention it might even be useful to allow cartesian axes to be inserted in the grid separately, like with a `row` or `column` attribute that plucks the appropriate `domain` from the grid. Like if you make a splom (using `xaxes` and `yaxes`) and want something special along the diagonal (box plots for example) - these may want to share a y axis with the grid but have a separate x axis (categorical in that case). We might be able to do this with `overlaying` but that might not work because the axis it's overlaying doesn't actually have a subplot on the diagonal... anyway that seems a kind of hacky solution, would be cleaner to just specify the column.

Reply:
... and also allowing multiple grids could be a nice way to generate grids with [multi-colspan](https://github.com/plotly/plotly.py/blob/master/plotly/tests/test_core/test_tools/test_make_subplots.py#L849-L852), multi-rowspan cells and crazy shared axes [pattern](https://github.com/plotly/plotly.py/blob/cb71b64e22c29e91d11862c32dd533c2442b1d1f/plotly/tests/test_core/test_tools/test_make_subplots.py#L1627-L1629).

Reply:
I think multiple overlaying grids is likely to be more confusing than helpful. colspan/rowspan would be easier to understand and use as just that - a way to specify colspan and rowspan within the existing grid, some of them wouldn't even work as a different grid, unless you monkey with `grid.domain`, which would be super confusing, and getting the gaps to line up correctly would be automatic with colspan/rowspan but nontrivial between grids having a different number of rows and columns.

I suppose for my other use case, inserting extra subplots in an otherwise fully coupled layout, we could support providing `xaxes`, `yaxes`, *and* `subplots`.

> we could probably do `pie` (i.e. first pie trace in `gd.data`), `pie2` (2nd pie trace in `gd.data`), `pie3` and similar for other traces that provide their own subplot. But yeah, it's not ideal.

Yeah, my concern with that is order in `gd.data` can be changed, or you can delete or change the type of an earlier trace, and then you'd have to know to update the reference in the grid. Also it's ambiguous, does `pie2` mean the pie in trace 2 or the second pie?

Anyway, it's pretty low-footprint, if we do think of a nice way to do this I don't think it would be a problem to add it and continue supporting the way I have it now.